pkg/operator/controller/volumepopulator.go

```diff
@@ -0,0 +1,122 @@
+Copyright 2025 The CDI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	popv1beta1 "github.com/kubernetes-csi/volume-data-source-validator/client/apis/volumepopulator/v1beta1"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	cdiv1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
+	forklift "kubevirt.io/containerized-data-importer-api/pkg/apis/forklift/v1beta1"
+	cc "kubevirt.io/containerized-data-importer/pkg/controller/common"
+	"kubevirt.io/containerized-data-importer/pkg/util"
+)
+
+var volumePopulatorSources = []runtime.Object{
+	&cdiv1.VolumeImportSource{},
+	&cdiv1.VolumeCloneSource{},
+	&cdiv1.VolumeUploadSource{},
+	&forklift.OpenstackVolumePopulator{},
+	&forklift.OvirtVolumePopulator{},
+}
+
+func ensureVolumePopulatorsExist(ctx context.Context, c client.Client, scheme *runtime.Scheme) (bool, error) {
+	cdi, err := cc.GetActiveCDI(ctx, c)
+	if err != nil {
+		return false, err
+	}
+	if cdi == nil {
+		return false, fmt.Errorf("no active CDI")
+	}
+	installerLabels := util.GetRecommendedInstallerLabelsFromCr(cdi)
+
+	for _, source := range volumePopulatorSources {
+		gvk, err := c.GroupVersionKindFor(source)
+		if err != nil {
+			return false, err
+		}
+		popName := strings.ToLower(gvk.Kind)
+		desiredPop := &popv1beta1.VolumePopulator{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: popName,
+			},
+			SourceKind: metav1.GroupKind(gvk.GroupKind()),
+		}
+		util.SetRecommendedLabels(desiredPop, installerLabels, "cdi-operator")
+
+		key := client.ObjectKey{Name: popName}
+
+		currentPop := &popv1beta1.VolumePopulator{}
+		err = c.Get(ctx, key, currentPop)
+		if err == nil {
+			if currentPop.SourceKind != metav1.GroupKind(gvk.GroupKind()) {
+				currentPop.SourceKind = metav1.GroupKind(gvk.GroupKind())
+				if err := c.Update(ctx, currentPop); err != nil {
+					return false, err
+				}
+			}
+			continue
+		}
+
+		if !errors.IsNotFound(err) {
+			return false, err
+		}
+
+		if err = controllerutil.SetControllerReference(cdi, desiredPop, scheme); err != nil {
+			return false, err
+		}
+
+		if err := c.Create(ctx, desiredPop); err != nil {
+			return false, err
+		}
+	}
+	return true, nil
```

Comment:
PR looks great only thing I was worried about is event spam over updated==true,
which turned out to be the case:
```bash
$ k get events --all-namespaces -o json | jq -r '.items[] | select(.message == "Successfully ensured volume populators exist") | .count'
8
```
(grows over time, this gets caught in conformance tests in openshift for example)

Here's a reference for a similar issue:
https://github.com/kubevirt/containerized-data-importer/pull/3490